⚡ Fix GRPO PEFT

qgallouedec opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly. They may suggest changes to make the code even better.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

tgaddair commented on trl/trainer/grpo_trainer.py:
```diff
@@ -388,7 +389,11 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
-                    state_dict = unwrapped_model.state_dict()
+                    if isinstance(unwrapped_model, PeftModel):
+                        unwrapped_model = copy.deepcopy(unwrapped_model)
+                        unwrapped_model.merge_and_unload()
```
Hey @qgallouedec, just want to call out that this is quite a compute and memory intensive operation.

Assuming the underlying Peft model is LoRA, vLLM already supports native dynamic LoRA inference, which can be done like so:

```
self.llm = LLM(
            ...,
            enable_lora=True,
        )

lora_request = LoRARequest(
     "<unique_str>",
     self.state.global_step,  # needs to be a unique identifier
     lora_path=checkpoint_dir_containing_adapter_weights,
)

outputs = self.llm.generate(
    all_prompts_text,
    sampling_params=self.sampling_params,
    lora_request=lora_request,
    use_tqdm=False
)
```

gagan3012 commented:
Hello, 
Are there any updates on this PR as this is currently a blocker in my work.

qgallouedec commented:
It's WIP, high priority. But if you find a fix in the meantime feel free to controbute 😊

tgaddair commented:
Hey @qgallouedec, please see #2730 for a follow-up that implements my suggestion.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_2725). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan commented on trl/trainer/grpo_trainer.py:
```diff
@@ -450,6 +450,20 @@ def _move_model_to_vllm(self):
+                    for k, v in state_dict.items()
+                    if self.model.prefix not in k
+                }
+                state_dict = {
+                    k.replace("modules_to_save.default.", ""): v
+                    for k, v in state_dict.items()
+                    if "original_module" not in k
+                }
```
Good catch to think about the `modules_to_save`. Note here that the adapter name can be different from `"default"`. You could get the adapter name from `model.active_adapters`, which is a list of all active adapters. I assume in this context, there can only ever be one (raise an error when more?), so taking the first item should work.

qgallouedec replied:
thanks for reviewing @BenjaminBossan! I'll make it multi-adaptater-proof, probably simpler.

BenjaminBossan commented on trl/trainer/grpo_trainer.py:
```diff
@@ -450,6 +450,20 @@ def _move_model_to_vllm(self):
+                unwrapped_model.merge_adapter()
+                state_dict = unwrapped_model.state_dict()
+                unwrapped_model.unmerge_adapter()
+                state_dict = {
```
It could be helpful to add a comment here on how the state dict is changed and why.

BenjaminBossan replied:
Maybe something like:

"Note that merge_adapter merges leaves the PEFT model in place, unlike merge_and_unload, but does not require a full copy. Therefore, we use this method, but the PEFT-specific prefixes need to be removed. Since the adapter weights are already merged into the base weights, we can remove those too. Next, adjust the prefix of modules_to_save (if present) and remove the corresponding original_module. After the state_dict creation, we can call unmerge_adapter to restore the PEFT model to its original state."

It's a bit verbose, not sure if that's an issue, but I prefer to give too much info rather than too little.

BenjaminBossan reviewed (commented):
Nice PR, LGTM. Just a minor comment about possible non-default adapter names.

albertsokol commented:
I've been trying out this PR with Qwen 2.5 14B, like so:

```python
model = AutoModelForCausalLM.from_pretrained(
    args.identifier,
    device_map="auto",
    torch_dtype=torch.bfloat16,
    attn_implementation="flash_attention_2",
)
lora_config = LoraConfig(...)

model = get_peft_model(model, lora_config)

trainer = GRPOTrainer(
    model=model,
    ...
    peft_config=lora_config if args.use_lora else None,
)
trainer.train()
```

But I get an error like this:
```
[rank0]: Traceback (most recent call last):
[rank0]:   File "/home/alberts/dialog/dialog/reward/train.py", line 176, in <module>
[rank0]:     main()
[rank0]:   File "/home/alberts/dialog/dialog/reward/train.py", line 170, in main
[rank0]:     trainer.train()
[rank0]:   File "/home/alberts/venv/lib/python3.10/site-packages/transformers/trainer.py", line 2171, in train
[rank0]:     return inner_training_loop(
[rank0]:   File "/home/alberts/venv/lib/python3.10/site-packages/transformers/trainer.py", line 2531, in _inner_training_loop
[rank0]:     tr_loss_step = self.training_step(model, inputs, num_items_in_batch)
[rank0]:   File "/home/alberts/venv/lib/python3.10/site-packages/transformers/trainer.py", line 3669, in training_step
[rank0]:     inputs = self._prepare_inputs(inputs)
[rank0]:   File "/home/alberts/venv/lib/python3.10/site-packages/trl/trainer/grpo_trainer.py", line 498, in _prepare_inputs
[rank0]:     llm_model.load_weights(state_dict.items())
[rank0]:   File "/home/alberts/venv/lib/python3.10/site-packages/vllm/model_executor/models/qwen2.py", line 515, in load_weights
[rank0]:     return loader.load_weights(weights)
[rank0]:   File "/home/alberts/venv/lib/python3.10/site-packages/vllm/model_executor/models/utils.py", line 235, in load_weights
[rank0]:     autoloaded_weights = set(self._load_module("", self.module, weights))
[rank0]:   File "/home/alberts/venv/lib/python3.10/site-packages/vllm/model_executor/models/utils.py", line 224, in _load_module
[rank0]:     raise ValueError(msg)
[rank0]: ValueError: There is no module or parameter named 'base_model' in Qwen2ForCausalLM
```

Any idea what's going on there? Is it the case that Qwen 2 doesn't have the `base_model` key or something like that? The model type that's getting passed to my Trainer is `<class 'peft.peft_model.PeftModelForCausalLM'>` so I think it should be correct...

qgallouedec commented:
Thanks for reporting, I'll try to reproduce

winglian commented:
> [rank0]: ValueError: There is no module or parameter named 'base_model' in Qwen2ForCausalLM
> ```
> 
> Any idea what's going on there? Is it the case that Qwen 2 doesn't have the `base_model` key or something like that? The model type that's getting passed to my Trainer is `<class 'peft.peft_model.PeftModelForCausalLM'>` so I think it should be correct...

@qgallouedec In the PR I have up, I had to do https://github.com/huggingface/trl/pull/2818/files#diff-964e6fd373aa93037604064cb2b822d7f8e2735e33f791065acf2c4c3552d393R457-R458

qgallouedec commented:
We were not able to reproduce the error. Do you have a code that would trigger this error?

winglian commented on tests/test_grpo_trainer.py:
```diff
@@ -498,3 +498,38 @@ def test_training_with_sync_ref_model(self):
+                report_to="none",
+            )
+            trainer = GRPOTrainer(
+                model="trl-internal-testing/small-Qwen2ForCausalLM-2.5",
```
This test didn't actually test PEFT/lora

albertsokol commented:
@winglian Thanks very much, yes I was able to confirm that when I use the PR by @winglian the training proceeds correctly.

albertsokol commented:
@qgallouedec as for my training code I'm running the following script with `python train.py`, I found that trying to use accelerate causes OOM in all my experiments, but with `python` I can now train 14B model with just 2x GPU. (However, only if I use cuda:0 for VLLM generation, otherwise, I get an error).

```
    model = AutoModelForCausalLM.from_pretrained(
        "Qwen-2.5-14B-Instruct",
        device_map="auto",
        torch_dtype=torch.bfloat16,
        attn_implementation="flash_attention_2",
    )

    dataset = create_huggingface_dataset(...)

    train_args = GRPOConfig(
        # Training hyperparameters
        num_train_epochs=10,
        per_device_train_batch_size=8,
        gradient_accumulation_steps=4,
        warmup_ratio=0.005,
        learning_rate=3e-6,
        # Logging and saving
        logging_dir=str(log_dir),
        report_to=["tensorboard"],
        logging_steps=5,
        eval_strategy="no",
        output_dir=str(output_dir),
        save_steps=50,
        # Device options
        gradient_checkpointing=False,
        bf16=True,
        use_vllm=True,
        vllm_gpu_memory_utilization=0.38,
        vllm_max_model_len=args.max_seq_length,
        vllm_device="cuda:0",
        # GRPO/generation hyperparameters
        num_generations=8,
        max_completion_length=args.max_completion_length,
        beta=0.04,
        temperature=0.7,
    )

    lora_config = LoraConfig(
        r=16,
        lora_alpha=32,
        lora_dropout=0.05,
        bias="none",
        task_type="CAUSAL_LM",
        init_lora_weights="gaussian",
        target_modules=[
            "q_proj",
            "k_proj",
            "v_proj",
            "o_proj",
            "gate_proj",
            "up_proj",
            "down_proj",
        ],
    )

    accelerator = Accelerator()
    reward_model_adapter = RewardModelAdapter(
        reward_model=reward_model,
        accelerator=accelerator,
        log_dir=log_dir,
    )


    model.gradient_checkpointing_enable()
    model.enable_input_require_grads()
    model = get_peft_model(model, lora_config)

    trainer = GRPOTrainer(
        model=model,
        args=train_args,
        train_dataset=dataset,
        reward_funcs=[reward_model_adapter],
        peft_config=lora_config if args.use_lora else None,
    )

    trainer.train()
```

Output of `trl env`:
```
- Platform: Linux-5.10.0-33-cloud-amd64-x86_64-with-glibc2.31
- Python version: 3.10.16
- PyTorch version: 2.5.1
- CUDA device(s): NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB
- Transformers version: 4.48.2
- Accelerate version: 1.3.0
- Accelerate config: not found
- Datasets version: 3.2.0
- HF Hub version: 0.28.1
- TRL version: 0.15.0.dev0 (from commit 81221661c6f864bd7cdb7c461e881bbe03414be8, the latest in `main`)
- bitsandbytes version: not installed
- DeepSpeed version: not installed
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.59.7
- PEFT version: 0.14.0
```

winglian commented:
here's a repro https://gist.github.com/winglian/b534238c44833b4478f7518ebd0f2598

AndreiCComan commented:
Hello, thanks to everyone who helped fix this and for all the effort!

I’ve come across a potential issue and was wondering if anyone else has experienced it. When `use_peft: true` is set, the behavior differs depending on `use_vllm`. If `use_vllm: false`, the model appears to learn correctly, with the reward curve displaying the expected pattern. However, when `use_vllm: true`, the reward curve remains mostly flat.

Any insights into what could cause this behavior?

qgallouedec commented:
> here's a repro https://gist.github.com/winglian/b534238c44833b4478f7518ebd0f2598

Thanks, I was able to reproduce, the error comes from

```diff
- get_peft_model(model, lora_config)
+ model = get_peft_model(model, lora_config)
```

I've added the missing test in your branch, added some comments for clarity, and I think we're good to merge.

TweedBeetle commented:
@albertsokol, would you perhaps be willing to share what max_seq_length you are working with?
